### PR TITLE
Reload config on change

### PR DESCRIPTION
**Pull request by margaale**
#525 

You mean something like this?

### CODE REVIEW

File: src/Ocelot/Middleware/OcelotMiddlewareExtensions.cs [hunk 1]
```diff
@@ -155,7 +155,16 @@ private static bool AdministrationApiInUse(IAdministrationPath adminPath)
         }
 
         private static async Task SetFileConfigInConsul(IApplicationBuilder builder,
-            IFileConfigurationRepository fileConfigRepo, IOptions<FileConfiguration> fileConfig,
+          IFileConfigurationRepository fileConfigRepo, IOptionsMonitor<FileConfiguration> fileConfig,
+          IInternalConfigurationCreator internalConfigCreator, IInternalConfigurationRepository internalConfigRepo)
+        {
+            fileConfig.OnChange(async (config) => await SetFileConfigInConsul(builder, fileConfigRepo, fileConfig, internalConfigCreator, internalConfigRepo));
```
Comment (TomPallister) [inline src/Ocelot/Middleware/OcelotMiddlewareExtensions.cs]: Actually I thought about this more and we shouldnt do it, the whole point is the config is in consul and it doesnt actually write it to disk.

**Comment by TomPallister** [comment]
@margaale this seems like a fine approach to me, and I guess people who dont want it to reload just have to pass false when they setup. Is it possible to add some tests for this? Probably at acceptance test level?

**Comment by margaale** [comment]
I've undone the consul change. Tomorrow i'll try adding the tests...

**Comment by TomPallister** [comment]
Awesome! Thanks for this.

**Comment by margaale** [comment]
No problem. It's useful for both of us. Sorry for the consul things. I actually did it for the sake of completeness. I don't know what is really... :-D

**Comment by TomPallister** [comment]
Haha all good!

**Comment by margaale** [comment]
Hi, i'm gonna need your help. I'm having trouble figuring out how should i write the test. Right now i have:

`   [Fact]
        public void should_reload_config_on_change()
        {
            var initialConfig = new FileConfiguration
            {
                GlobalConfiguration = new FileGlobalConfiguration
                {
                    RequestIdKey = "initialKey"
                }
            };

            var anotherConfig = new FileConfiguration
            {
                GlobalConfiguration = new FileGlobalConfiguration
                {
                    RequestIdKey = "someOtherKey"
                }
            };           
            
            this.Given(x => _steps.GivenThereIsAConfiguration(initialConfig))
                .And(x => _steps.GivenOcelotIsRunningReloadingConfig(true))
                .And(x => _steps.GivenThereIsAConfiguration(anotherConfig))
                .And(x => )
                .BDDfy();
        }`

That last and should check that the config has changed, but i don't know how to check for that without mocking some component. How would you do it?

**Comment by margaale** [comment]
I rebased my branch because i didn't like the way it was. Please, review it again, sorry. Also added a test. Probably  i should have added one for admin api and consul too.

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 2]
```diff
@@ -54,6 +55,18 @@ public Steps()
...
+        public async Task ThenConfigShouldBe(FileConfiguration fileConfig)
+        {
+            var internalConfigCreator = _ocelotServer.Host.Services.GetService<IInternalConfigurationCreator>();
+
+            var internalConfigRepo = _ocelotServer.Host.Services.GetService<IInternalConfigurationRepository>();
+            var internalConfig = internalConfigRepo.Get();
+
+            var config = await internalConfigCreator.Create(fileConfig);
+
+            internalConfig.Data.RequestId.ShouldBe(config.Data.RequestId);
```
Comment (margaale) [inline test/Ocelot.AcceptanceTests/Steps.cs]: I don't like this. But shouldly doesn't implement an ShouldBeEquivalentTo method as FluentAssertions does. Do you have any suggestion?

Comment (TomPallister) [reply test/Ocelot.AcceptanceTests/Steps.cs]: Dont worry this is fine, Yeh FluentAssertions has done well adding that feature...maybe I will move to them one day!

**Pull request merged**